src/aggregator/server/server.go

```diff
@@ -0,0 +1,192 @@
+	"github.com/m3db/m3/src/cmd/services/m3aggregator/serve"
+	xconfig "github.com/m3db/m3/src/x/config"
+	"github.com/m3db/m3/src/x/instrument"
+
+	"go.uber.org/zap"
+)
+
+const (
+	gracefulShutdownTimeout = 15 * time.Second
+)
+
+// RunOptions is the server options for running the aggregator server.
+type RunOptions struct {
+	Config config.Configuration
+}
+
+// Run runs the aggregator server.
+func Run(opts RunOptions) {
+	cfg := opts.Config
+
+	// Create logger and metrics scope.
+	logger, err := cfg.Logging.BuildLogger()
+	if err != nil {
+		// NB(r): Use fmt.Fprintf(os.Stderr, ...) to avoid etcd.SetGlobals()
+		// sending stdlib "log" to black hole. Don't remove unless with good reason.
+		fmt.Fprintf(os.Stderr, "error creating logger: %v\n", err)
+		os.Exit(1)
+	}
+	defer logger.Sync()
+
+	xconfig.WarnOnDeprecation(cfg, logger)
+
+	scope, closer, err := cfg.Metrics.NewRootScope()
+	if err != nil {
+		logger.Fatal("error creating metrics root scope", zap.Error(err))
+	}
+	defer closer.Close()
+	instrumentOpts := instrument.NewOptions().
+		SetLogger(logger).
+		SetMetricsScope(scope).
+		SetMetricsSamplingRate(cfg.Metrics.SampleRate()).
+		SetReportInterval(cfg.Metrics.ReportInterval())
+
+	buildReporter := instrument.NewBuildReporter(instrumentOpts)
+	if err := buildReporter.Start(); err != nil {
+		logger.Fatal("could not start build reporter", zap.Error(err))
+	}
+
+	defer buildReporter.Stop()
+
+	var (
+		m3msgAddr        string
+		m3msgServerOpts  m3msg.Options
+		rawTCPAddr       string
+		rawTCPServerOpts rawtcp.Options
+		httpAddr         string
+		httpServerOpts   http.Options
+	)
+	if cfg.M3Msg != nil {
+		// Create the M3Msg server options.
+		m3msgAddr = cfg.M3Msg.Server.ListenAddress
+		m3msgInsrumentOpts := instrumentOpts.
+			SetMetricsScope(scope.
+				SubScope("m3msg-server").
+				Tagged(map[string]string{"server": "m3msg"}))
+		m3msgServerOpts, err = cfg.M3Msg.NewServerOptions(m3msgInsrumentOpts)
+		if err != nil {
+			logger.Fatal("could not create m3msg server options", zap.Error(err))
+		}
+	}
+
+	if cfg.RawTCP != nil {
+		// Create the raw TCP server options.
+		rawTCPAddr = cfg.RawTCP.ListenAddress
+		rawTCPInstrumentOpts := instrumentOpts.
+			SetMetricsScope(scope.
+				SubScope("rawtcp-server").
+				Tagged(map[string]string{"server": "rawtcp"}))
+		rawTCPServerOpts = cfg.RawTCP.NewServerOptions(rawTCPInstrumentOpts)
+	}
+
+	if cfg.HTTP != nil {
+		// Create the http server options.
+		httpAddr = cfg.HTTP.ListenAddress
+		httpServerOpts = cfg.HTTP.NewServerOptions()
+	}
+
+	// Create the kv client.
+	client, err := cfg.KVClient.NewKVClient(instrumentOpts.
+		SetMetricsScope(scope.SubScope("kv-client")))
+	if err != nil {
+		logger.Fatal("error creating the kv client", zap.Error(err))
+	}
+
+	// Create the runtime options manager.
+	runtimeOptsManager := cfg.RuntimeOptions.NewRuntimeOptionsManager()
+
+	// Create the aggregator.
+	aggregatorOpts, err := cfg.Aggregator.NewAggregatorOptions(rawTCPAddr,
```

Comment:
Is it fine if `rawTCPAddr` is unset here?

Reply:
nvm saw this was copied over :+1:

Reply:
Yeah this is just copied 👍